File: 백준/Silver/15650. N과 M （2）/N과 M （2）.py
def solve(index,arr):

    #1. 기저조건: index ==M
    if index == M:
        result.append(arr)

        return
    # 2. 구현
    for i in range(1,N+1):
        # arr 배열에 첫번째로 들어가고,
        if len(arr)==0 and visited[i]==0:
            # 방문하지 않은 숫자라면 바로 추가
            visited[i]=1
            solve(index+1,arr+[i])
        # arr 배열에 첫번째가 아니고, 현재 arr의 마지막 숫자보다 작고, 방문한적 없으면
        if len(arr)>0 and arr[index-1] < i and visited[i]==0:
            # 추가 하고 재귀 돌기

            visited[i]=1
            solve(index+1,arr+[i])
        # 재귀 나와서 visited 취소하기
        visited[i]=0




N, M = map(int,input().split())
visited =[0]*(N+1)
result = []

solve(0,[])
for i in result:
    print(*i)

File: 백준/Silver/15650. N과 M （2）/N과 M （2）.java
import java.io.*;
import java.util.ArrayList;

public class Main {

    static int N, M;
    static ArrayList<Integer>numbers = new ArrayList<>();
    static boolean[] visited;
    static StringBuilder sb = new StringBuilder();
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String[] input = br.readLine().split(" ");

        N = Integer.parseInt(input[0]);
        M = Integer.parseInt(input[1]);
        visited = new boolean[N+1];
        backtrack(0);

        System.out.println(sb);

    }
    private static void backtrack (int depth){
        // 기저조건
        if(depth == M){
             //출력리스트에 추가
            for(int num : numbers){
                sb.append(num).append(' ');
            }
            sb.append('\n');
            return;
        }

        // 구현
        for(int i = 1 ; i <= N; i++){
            // 현재 배열에 아무 것도 없거나 마지막 숫자보다 현재 숫자가 더 크고
            if(numbers.size() == 0 || numbers.get(numbers.size()-1) < i ){
                //방문한 적 없으면 넣기
                if(!visited[i]){
                    numbers.add(i);
                    visited[i] = true;
                    backtrack(depth+1);
                    numbers.remove(numbers.size()-1);
                    visited[i] = false;
                }
            }
        }
    }
}
